### ISSUE TITLE
execution model

### ISSUE THREAD

**Issue by hawkw** [hawkw/mycelium#4]
or, "how is wasm formed? how user get executable?"

are we...

#### ...running a wasm VM in the kernel? 

+ if so, does it really make sense to talk about "user space" and "kernel space" any more? 
+ e.g. can we just run everything in ring 0 and have the "system calls" just be host fns the wasm VM calls into?
   - there's almost certainly a perf benefit from eliminating CPU context switches this way...
   - but there's _also_ almost certainly a perf disadvantage from "running everything in a VM"
   - what is the security model for this?
     * are we relying on the wasm VM for isolation? then we need to be able to trust it
     * or, do we use the wasm VM _and_ the CPU's privilege rings for defence-in-depth? then we have to do system calls via interrupts like "normal people" do
 + are we JITting code?

#### ...pre-compiling code w/ lucet?
  
+ if we do the compilation in the kernel and trust the lucet we have in the kernel, can we stil run the lucet-compiled code in ring 0 to avoid context switches?
   - is this a terrible idea
+ when does the compilation happen? do users invoke lucet or does the OS?
   - if the OS is responsible for compiling, does that mean we recompile a wasm into a "real binary" every time a program runs? that seems Bad?
   - does the kernel cache compiled "real binaries"? 
      * if we rely on the wasm model for isolation, we would need to verify that the cache wasn't tampered with? how do we do that?
      * the kernel could sign code it compiles?
      * or we could implement the caching in a way that is inaccessible to userspace...but then you could tamper with the block storage device under a different OS.
      * maybe the kernel signs the whole cache somehow?
      * there's probably "some kind of SGX bullshit" we could use for this, maybe?

**Comment by hawkw** [comment]
So, a thought: it definitely seems like having the kernel manage AOT compilation of wasms is the correct approach. However, it also has perhaps the most moving parts, and introduces some more subsystems (managing the compiled artefact cache + ensuring its integrity). 

Would it be worthwhile to consider starting with a wasm interpreter to run wasms & adding in AOT compilation transparently? That way, we can start "running code" sooner & iterate on the userland as well. I guess my question is whether this would make our lives easier or _harder_?

**Comment by mystor** [comment]
A well-designed interface should make it easy-enough to switch between different wasm backends. It should be reasonable to start with an interpreter, and move on from there. That should help with getting pieces put together before we get AOT compilation, caching, etc. fully set-up.

**Comment by hawkw** [comment]
Yeah, and there may also be a use-case for both, as well — e.g., if I'm running mycelium and I want to, say, test some code I'm writing, I might want to take the perf hit of the wasm interpreter in exchange for not having to spend extra time compiling my wasm into realcode. On the other hand, when the OS is fully featured enough to support doing software development on it (read: "nebulous future"), that wasm interpreter could also be a userspace thing...

**Comment by iximeow** [comment]
> However, it also has perhaps the most moving parts, and introduces some more subsystems (managing the compiled artefact cache + ensuring its integrity).

we can totally `just` AOT every time we load a wasm and worry about caching later. the "AOT" bit can just write to memory and we can copy that somewhere appropriate in the kernel and run it

this reminds me that lucet has some opinions about relocations (namely that they can be applied) so maybe using wasmtime to interpret modules would be easier to get running.

i have no idea what the landscape of wasm interpreters is, whatever's there it probably would involve even fewer moving pieces to start with.

between AOT or JIT, we can probably have the same interface from the kernel, so :woman_shrugging:

**Comment by hawkw** [comment]
> between AOT or JIT, we can probably have the same interface from the kernel, so 🤷‍♀

yeah, that was what I thought, we can swap the execution model out transparently. 

> we can totally `just` AOT every time we load a wasm and worry about caching later. the "AOT" bit can just write to memory and we can copy that somewhere appropriate in the kernel and run it

👍

**Comment by mystor** [comment]
The simplest wasm interpreter I know of is probably [watt's](https://github.com/dtolnay/watt/tree/fd8e5676b2b1fe478bcc7a90ff11fe8d6c04d466/runtime). It's pretty allocation-heavy, has a weird API, and contains no `unsafe` code. If it's easy enough to use a proper interpreter like wasmtime's, we should go for it, but we might be able to hack something small together with watt's interpreter relatively easily. I think we'd only need `Vec` (+ `String`), `Rc`, and (maybe) `HashMap`.
